### ISSUE TITLE
`addNewDrawnMutation( )` improperly vectorized

### ISSUE THREAD

**Issue by petrelharp** [MesserLab/SLiM#64]
As pointed out over [here](https://github.com/tskit-dev/pyslim/issues/59), there seems to be an issue when `addNewDrawnMutation()` is called with a vector of positions. A haphazard collection of the mutations after the first mutation are recorded in the MutationTable without metadata. I haven't checked if this is a problem only with tree sequence recording or if the actual mutations are wrong in SLiM.

Here's the MWE:
```
initialize() {
    // problem only happens about half the time
    setSeed(42);
	initializeTreeSeq();
	initializeMutationRate(0);
    initializeMutationType("m1", 0.5, "f", 0.1);
    defineConstant("chr_len", 10);
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, chr_len-1);
    initializeRecombinationRate(0);
}

1 {
	sim.addSubpop("p1", 2);
}

1 late() {
	target = sample(sim.subpopulations.genomes, 1);
    balancing_pos = sample(0:(chr_len-1), 2);
    // THIS does not work:
	target.addNewDrawnMutation(m1, asInteger(balancing_pos));

    // THIS works:
    // for (pos in balancing_pos) {
    //     target.addNewDrawnMutation(m1, pos);
    // }

    sim.treeSeqOutput("balancing_output.trees");
    sim.simulationFinished();
}
```
and python:
```
import pyslim

ts = pyslim.load("balancing_output.trees")

print(ts.tables.mutations)

g = ts.genotype_matrix()
```
The genotype matrix call throws the error, but the problem is visible in the mutation table (as the second mutation has no derived state and no metadata); here is the output:
```
// RunInitializeCallbacks():
initializeTreeSeq();
initializeMutationRate(0);
initializeMutationType(1, 0.5, "f", 0.1);
initializeGenomicElementType(1, m1, 1);
initializeGenomicElement(g1, 0, 9);
initializeRecombinationRate(0);

// Starting run at generation <start>:
1 

id	site	node	derived_state	parent	metadata
0	0	4	1	-1	AQAAAM3MzD0BAAAAAQAAAP8=
1	1	4		-1	
Traceback (most recent call last):
  File "debug_mutate_to_vcf.py", line 7, in <module>
    g = ts.genotype_matrix()
  File "/home/peter/.local/lib/python3.6/site-packages/tskit-0.2.4.dev0-py3.6-linux-x86_64.egg/tskit/trees.py", line 3035, in genotype_matrix
    impute_missing_data=impute_missing_data)
_tskit.LibraryError: Inconsistent mutations: state already equal to derived state
```

Thanks to @josieparis for reporting this problem.

**Comment by bhaller** [comment]
A quick look-over of the code didn't reveal anything obviously wrong, so this will require more thorough delving.  Stay tuned.

**Comment by petrelharp** [comment]
This appears to be a problem in tree sequence recording and not about SLiM's internal state. Furthermore, trying to reload a tree sequence file with this problem results in the `Inconsistent mutations` error, so there is no chance of invisible errors in SLiM simulation; only in analyzing which of the output mutations are SLiM mutations or not.

**Comment by bhaller** [comment]
Interesting.  I'm not sure whether/how/why the recorded TS information would be any different between the two patterns of Eidos calls, though, so the bug itself does presumably reside in SLiM's code...

**Comment by petrelharp** [comment]
After some printf debugging...
- the problem is that `derived_mutation_ids_at_position( )` is returning the wrong thing sometimes
- that is because this method assumes that the mutations are in order by position in the thing they're looking in
- and so, deleteing the two lines starting [here](https://github.com/MesserLab/SLiM/blob/dc1fee07905919dd0a2070968f0f92b9ad43d258/core/mutation_run.cpp#L277) fixes the error

The problem seems to occur because the mutations we are adding go to a single genome *not* in order by position (position 9 comes before position 2). Changing the recipe to do
```
    balancing_pos = sort(sample(0:(chr_len-1), 2));
```
also fixes the problem.

I don't understand the ramifications here, so I'm not proposing a fix.

**Comment by bhaller** [comment]
Nice debugging.  Yes, the sort order of mutations in the genomes gets corrupted, and that is an error in SLiM's internal state, not just in the tree-seq stuff.  I think I just neglected to consider that the positions given to addNewDrawnMutation() might not be in sorted order, and that case didn't get tested somehow.  I have modified the addNewDrawnMutation() code to sort the mutations before adding them, when necessary.
